src/hotspot/share/services/heapDumper.cpp

```diff
@@ -1598,7 +1598,7 @@ class JNILocalsDumper : public OopClosure {
 
 void JNILocalsDumper::do_oop(oop* obj_p) {
   // ignore null handles
-  oop o = *obj_p;
+  oop o = NativeAccess<AS_NO_KEEPALIVE>::oop_load(obj_p);
```

Comment:
The JNI Local roots do not need a load barrier, only JNI Global roots need that. The JNI Local roots are processed on safepoint entry as part of the "thread head" (via `ZStackWatermark::ZStackWatermark::start_processing_impl()`-> `Thread::oops_do_no_frames()`), so once you are in `VM_HeapDumper::do_thread()` the JNI Local roots have already passed a load barrier.

Reply:
Get it, Thanks!